integrations/google_genai/src/haystack_integrations/components/generators/google_genai/chat/chat_generator.py

```diff
@@ -359,37 +374,65 @@ def from_dict(cls, data: Dict[str, Any]) -> "GoogleGenAIChatGenerator":
         return default_from_dict(cls, data)
 
     def _process_streaming_chunk(
-        self, chunk: types.GenerateContentResponse, all_text_parts: List[str], all_tool_calls: List[ToolCall]
-    ) -> str:
+        self,
+        chunk: types.GenerateContentResponse,
+        index: int,
+        all_text_parts: List[str],
+        all_tool_calls: List[ToolCall],
+    ) -> StreamingChunk:
         """
         Process a single streaming chunk and extract text and tool calls.
 
         :param chunk: The streaming chunk from Google Gen AI.
+        :param index: The index of the chunk.
         :param all_text_parts: List to collect text parts.
         :param all_tool_calls: List to collect tool calls.
-        :returns: The text content from this chunk.
+        :returns: A single StreamingChunk object for this chunk.
         """
-        chunk_text = ""
+        content = ""
+        tool_calls = []
+        finish_reason = None
 
-        # Process candidates in the chunk
         if chunk.candidates:
-            candidate = chunk.candidates[0]  # Take the first candidate
+            candidate = chunk.candidates[0]
+            finish_reason = getattr(candidate, "finish_reason", None)
 
-            # Extract content from parts
             if candidate.content and candidate.content.parts:
+                tc_index = -1
                 for part in candidate.content.parts:
                     if part.text:
-                        chunk_text += part.text
                         all_text_parts.append(part.text)
+                        content += part.text
+
                     elif part.function_call:
+                        tc_index += 1
                         tool_call = ToolCall(
                             tool_name=part.function_call.name or "",
-                            arguments=dict(part.function_call.args) if part.function_call.args else {},
+                            arguments=part.function_call.args if part.function_call.args else {},
                             id=part.function_call.id,
                         )
                         all_tool_calls.append(tool_call)
+                        tool_calls.append(
+                            ToolCallDelta(
+                                # Google GenAI does not provide index, but it is required for tool calls
+                                index=tc_index,
+                                id=tool_call.id,
+                                tool_name=tool_call.tool_name,
+                                arguments=json.dumps(tool_call.arguments) if tool_call.arguments else None,
```

Comment:
okay interesting you needed to use `json.dumps` here because Google returns a dict already for arguments?

Reply:
yes

```python
class FunctionCall(_common.BaseModel):
  """A function call."""
...
  args: Optional[dict[str, Any]] = Field(
      default=None,
      description="""Optional. The function parameters and values in JSON object format. See [FunctionDeclaration.parameters] for parameter details.""",
  )
```